packages/editor/src/components/post-actions/set-as-homepage.js

```diff
@@ -0,0 +1,217 @@
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { useEffect, useMemo, useState } from '@wordpress/element';
+import {
+	Button,
+	__experimentalText as Text,
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+} from '@wordpress/components';
+import { useDispatch, useSelect, resolveSelect } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import { store as noticesStore } from '@wordpress/notices';
+
+/**
+ * Internal dependencies
+ */
+import { getItemTitle } from '../../utils/get-item-title';
+import { unlock } from '../../lock-unlock';
+
+const SetAsHomepageModal = ( { items, closeModal, onActionPerformed } ) => {
+	const [ item ] = items;
+	const pageTitle = getItemTitle( item );
+	const { showOnFront, currentHomePage } = useSelect( ( select ) => {
+		const { getEntityRecord } = select( coreStore );
+		const { getHomePage } = unlock( select( coreStore ) );
+		const siteSettings = getEntityRecord( 'root', 'site' );
+		const homePage = getHomePage();
+		return {
+			showOnFront: siteSettings?.show_on_front,
+			currentHomePage: homePage.postId,
+		};
+	} );
+	const [ currentHomePageTitle, setCurrentHomePageTitle ] = useState( '' );
+	const isPageDraft = item.status === 'draft';
+
+	const { getEntityRecord } = resolveSelect( coreStore );
+	const { saveEditedEntityRecord, saveEntityRecord } =
+		useDispatch( coreStore );
+	const { createSuccessNotice, createErrorNotice } =
+		useDispatch( noticesStore );
+
+	useEffect( () => {
+		const handleGetCurrentHomepage = async () => {
+			const currentHomePageItem = await getEntityRecord(
+				'postType',
+				'page',
+				currentHomePage
+			);
+			setCurrentHomePageTitle( getItemTitle( currentHomePageItem ) );
+		};
+		handleGetCurrentHomepage();
+	}, [ currentHomePage, getEntityRecord ] );
+
+	async function onSetPageAsHomepage( event ) {
+		event.preventDefault();
+
+		try {
+			// If selected page is set to draft, publish the page.
+			if ( isPageDraft ) {
+				await saveEntityRecord( 'postType', 'page', {
+					...item,
+					status: 'publish',
+				} );
+			}
+
+			// Save new home page settings.
+			await saveEntityRecord( 'root', 'site', {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			closeModal?.();
+
+			await saveEditedEntityRecord( 'root', 'site', undefined, {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
```

Comment:
```suggestion
			await saveEditedEntityRecord( 'root', 'site', undefined, {
				page_on_front: item.id,
				show_on_front: 'page',
			} );
			
			closeModal?.();
```

edited entity record calls `saveEntityRecord` under the hood should it might be possible to omit.

Reply:
We should only require this but there may be a bug here that means we need both to correctly persist.

Reply:
Just confirming that I tested using only `saveEditedEntityRecord` rather than both `saveEntityRecord` and `saveEditedEntityRecord`, and the reading settings did not update successfully. I'm hoping we can address this in a follow-up, I can open a separate issue.

Reply:
Opened an issue for this here: https://github.com/WordPress/gutenberg/issues/67161.